apps/rush-lib/src/logic/pnpm/PnpmProjectShrinkwrapFile.ts

```diff
@@ -0,0 +1,238 @@
+
+    return projectShrinkwrapMap;
+  }
+
+  private _addDependencyRecursive(
+    projectShrinkwrapMap: Map<string, string>,
+    name: string,
+    version: string,
+    parentShrinkwrapEntry: IPnpmShrinkwrapDependencyYaml,
+    throwIfShrinkwrapEntryMissing: boolean = true
+  ): void {
+    const shrinkwrapEntry: IPnpmShrinkwrapDependencyYaml | undefined = this.shrinkwrapFile.getShrinkwrapEntry(
+      name,
+      version
+    );
+
+    if (!shrinkwrapEntry) {
+      if (throwIfShrinkwrapEntryMissing) {
+        throw new InternalError(`Unable to find dependency ${name} with version ${version} in shrinkwrap.`);
+      }
+      return;
+    }
+
+    const specifier: string = `${name}@${version}`;
+    let integrity: string | undefined = shrinkwrapEntry?.resolution?.integrity;
+    if (!integrity) {
+      // git dependency specifiers do not have an integrity entry. Instead, they specify the tarball field.
+      // So instead, we will hash the contents of the dependency entry and use that as the integrity hash.
+      // Ex:
+      // github.com/chfritz/node-xmlrpc/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38:
+      //   ...
+      //   resolution:
+      //     tarball: 'https://codeload.github.com/chfritz/node-xmlrpc/tar.gz/948db2fbd0260e5d56ed5ba58df0f5b6599bbe38'
+      const sha256Digest: string = crypto
+        .createHash('sha256')
+        .update(JSON.stringify(shrinkwrapEntry))
+        .digest('hex');
+      integrity = `${name}@${version}:${sha256Digest}:`;
+    }
+
+    const existingSpecifier: string | undefined = projectShrinkwrapMap.get(specifier);
+    if (existingSpecifier) {
+      if (existingSpecifier !== integrity) {
+        throw new Error(`Collision: ${specifier} already exists in with a different integrity`);
+      }
+      return;
+    }
+
+    // Add the current dependency
+    projectShrinkwrapMap.set(specifier, integrity);
+
+    // Add the dependencies of the dependency
+    for (const [name, version] of Object.entries(shrinkwrapEntry.dependencies || {})) {
+      this._addDependencyRecursive(projectShrinkwrapMap, name, version, shrinkwrapEntry);
+    }
+
+    // Add the optional dependencies of the dependency, and don't blow up if they don't exist
+    for (const [name, version] of Object.entries(shrinkwrapEntry.optionalDependencies || {})) {
+      this._addDependencyRecursive(
+        projectShrinkwrapMap,
+        name,
+        version,
+        shrinkwrapEntry,
+        (throwIfShrinkwrapEntryMissing = false)
+      );
+    }
+
+    // When using workspaces, hoisting of peer dependencies to a singular top-level project is not possible.
+    // Therefore, all packages that are consumed should be specified in the dependency tree. Given this, there
+    // is no need to look for peer dependencies, since it is simply a constraint to be validated by the
+    // package manager.
+    if (!this.shrinkwrapFile.isWorkspaceCompatible()) {
```

Comment:
This should probably just get passed into this function instead of calling this over and over.

Reply:
Made this a field so now it's no overhead :)